### ISSUE TITLE
Inconsistent behaviour of `set_slice`

### ISSUE THREAD

**Issue by dalcde** [hoodmane/rust_ext#23]
In `Matrix`, the `set_slice` function is idempotent, but not in `FpVector`.

**Comment by hoodmane** [comment]
Yeah, so the `Matrix` version doesn't nest correctly but the `FpVector` version does. The reason the behavior is different is because the `Matrix` set_slice never occurs nested (in fact is hardly used at all) whereas the `FpVector` version appears nested and needs proper scoping.

The ideal thing to do would be to make slicing a vector returns a wrapper which mutably borrows the original vector and implements `FpVectorT`. This would nest correctly without the wordy get_slice set_slice modify restore_slice approach I currently use. This is what it did in the C code but it's harder to implement in Rust and what we have is good enough so I was too lazy. Anyways, failing that it would be an improvement to make `Matrix.set_slice` mirror the behavior of `FpVector.set_slice`.

**Comment by dalcde** [comment]
I think the current usage of `Matrix.set_slice` is consistent with it nesting, which is good

**Comment by hoodmane** [comment]
Doesn't it break if you nest it? At least I thought it would, I haven't looked carefully.

**Comment by dalcde** [comment]
I haven't tried but I think we clear the slice before slicing it again
